java/BUILD.bazel

```diff
@@ -143,52 +145,62 @@ filegroup(
     ],
 )
 
-# plasma_store_server binary should be located in jar root path
-# and keep in sync with mvn resource
-copy_file(
-    name = "cp_plasma_store_server",
-    src = "//:plasma_store_server",
-    out = "plasma_store_server",
-)
-
-# `//:redis-server`'s full path is `external/com_github_antirez_redis/redis-server`,
-# This rule removes the prefix, and only keeps `redis-server`.
-copy_file(
-    name = "redis-server-without-prefix",
-    src = "//:redis-server",
-    out = "redis-server",
-)
+native_java_binary("runtime", "raylet", "//:raylet")
+native_java_binary("runtime", "plasma_store_server", "//:plasma_store_server")
+native_java_binary("runtime", "redis-server", "//:redis-server")
+native_java_binary("runtime", "gcs_server", "//:gcs_server")
+native_java_binary("runtime", "libray_redis_module.so", "//:libray_redis_module.so")
+native_java_library("runtime", "core_worker_library_java", "//:libcore_worker_library_java.so")
 
 filegroup(
     name = "java_native_deps",
     srcs = [
-        ":cp_plasma_store_server",
-        ":redis-server-without-prefix",
-        "//:core_worker_library_java",
-        "//:gcs_server",
-        "//:libray_redis_module.so",
-        "//:raylet",
+        ":raylet",
+        ":plasma_store_server",
+        ":redis-server",
+        ":gcs_server",
+        ":libray_redis_module.so",
+        ":core_worker_library_java",
     ],
 )
 
 # Generates the dependencies needed by maven.
 genrule(
-    name = "gen_maven_deps",
+    name = "cp_java_generated",
     srcs = [
         ":all_java_proto",
-        ":java_native_deps",
         ":copy_pom_file",
     ],
-    outs = ["gen_maven_deps.out"],
+    outs = ["cp_java_generated.out"],
     cmd = """
         WORK_DIR="$$(pwd)"
         # Copy protobuf-generated files.
         rm -rf "$$WORK_DIR/java/runtime/src/main/java/io/ray/runtime/generated"
         for f in $(locations //java:all_java_proto); do
             unzip "$$f" -x META-INF/MANIFEST.MF -d "$$WORK_DIR/java/runtime/src/main/java"
         done
+        date > $@
+    """,
+    local = 1,
+    tags = ["no-cache"],
+)
+
+# Generates the dependencies needed by maven.
+genrule(
+    name = "gen_maven_deps",
+    srcs = [
+        ":cp_java_generated",
+        ":java_native_deps",
+    ],
+    outs = ["gen_maven_deps.out"],
+    cmd = """
+        WORK_DIR="$$(pwd)"
         # Copy native dependecies.
-        NATIVE_DEPS_DIR="$$WORK_DIR/java/runtime/native_dependencies/"
+        OS_NAME="linux"
+        if [[ "$$OSTYPE" =~ darwin* ]]; then
```

Comment:
Please handle Windows everywhere across your PR (even if it it's just to produce an error in that case). Prefer to use `case` over `if [[ ]]`:

```
case "$${OSTYPE}" in
  linux*) ...;;
  darwin*) ...;;
  msys) ...;;
esac
```

Reply:
produced an error in that case